internal/resolution/entitysources/catalogdsource.go

```diff
@@ -0,0 +1,159 @@
+	panic("not implemented")
+}
+
+func (es catalogdEntitySource) Filter(ctx context.Context, filter input.Predicate) (input.EntityList, error) {
+	resultSet := input.EntityList{}
+	entities, err := getEntities(ctx, es.client)
+	if err != nil {
+		return nil, err
+	}
+	for _, entity := range entities {
+		if filter(&entity) {
+			resultSet = append(resultSet, entity)
+		}
+	}
+	return resultSet, nil
+}
+
+func (es catalogdEntitySource) GroupBy(ctx context.Context, fn input.GroupByFunction) (input.EntityListMap, error) {
+	entities, err := getEntities(ctx, es.client)
+	if err != nil {
+		return nil, err
+	}
+	resultSet := input.EntityListMap{}
+	for _, entity := range entities {
+		keys := fn(&entity)
+		for _, key := range keys {
+			resultSet[key] = append(resultSet[key], entity)
+		}
+	}
+	return resultSet, nil
+}
+
+func (es catalogdEntitySource) Iterate(ctx context.Context, fn input.IteratorFunction) error {
+	entities, err := getEntities(ctx, es.client)
+	if err != nil {
+		return err
+	}
+	for _, entity := range entities {
+		if err := fn(&entity); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func jsonMarshal(p interface{}) ([]byte, error) {
+	buf := &bytes.Buffer{}
+	dec := json.NewEncoder(buf)
+	dec.SetEscapeHTML(false)
+	err := dec.Encode(p)
+	if err != nil {
+		return nil, err
+	}
+	out := &bytes.Buffer{}
+	if err := json.Compact(out, buf.Bytes()); err != nil {
+		return nil, err
+	}
+	return out.Bytes(), nil
+}
+
+func getEntities(ctx context.Context, client client.Client) (input.EntityList, error) {
+	entities := input.EntityList{}
+	bundleMetadatas, packageMetdatas, err := fetchMetadata(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	for _, bundle := range bundleMetadatas.Items {
+		props := map[string]string{}
+
+		for _, prop := range bundle.Spec.Properties {
+			switch prop.Type {
+			case property.TypePackage:
+				type pkg struct {
+					Name    string `json:"packageName"`
+					Version string `json:"version"`
+				}
+				tmp := pkg{}
+				if err := json.Unmarshal(prop.Value, &tmp); err != nil {
+					return nil, err
+				}
+				pkgValue, err := jsonMarshal(tmp)
+				if err != nil {
+					return nil, err
+				}
+				props[property.TypePackage] = string(pkgValue)
+			}
+		}
+
+		imgValue, err := jsonMarshal(bundle.Spec.Image)
+		if err != nil {
+			return nil, err
+		}
+		props["olm.bundle.path"] = string(imgValue)
+		bundlePkg := packageMetdatas[bundle.Spec.Package]
+		channelName := ""
+		for _, ch := range bundlePkg.Spec.Channels {
+			for _, b := range ch.Entries {
+				if b.Name == bundle.Name {
+					channelName = ch.Name
+				}
+			}
+		}
+		channelValue, _ := jsonMarshal(property.Channel{ChannelName: channelName, Priority: 0})
+		props[property.TypeChannel] = string(channelValue)
+		entity := input.Entity{
+			ID:         deppy.IdentifierFromString(fmt.Sprintf("%s%s%s", bundle.Name, bundle.Spec.Package, channelName)),
+			Properties: props,
+		}
+		entities = append(entities, entity)
```

Comment:
This should go inside the channel loop.

For a given bundle, deppy expects the entity source to return an entity for each channel that the bundle is in (under a different ID). We can't just pick one of the channels and call it a day. So if a bundle is an entry in three channels, that should produce three entities.

Reply:
Oh yea that is a very good catch. Updating.